Replace `GoogleAnalyticsTracker` with Snowplow-backed `MeltanoTracker`

WillDaSilva opened the pull request:
Closes #5941
Closes #5959

edgarrmondragon committed:
Add general context schema

WillDaSilva committed:
Merge branch 'main' into drop-ga-tracker

edgarrmondragon committed:
Create general Snowplow context for Meltano events

WillDaSilva committed:
Merge branch 'drop-ga-tracker' of github.com:meltano/meltano into drop-ga-tracker

WillDaSilva committed:
Sort dependencies & add `backports-cached-property` and `tzlocal`

WillDaSilva committed:
Add return type hints for path-obtaining methods

WillDaSilva committed:
Add initial Snowplow schemas for command events, the environment context, and the plugin context

WillDaSilva committed:
Add `EnvironmentContext` class

This class can be instantiated with a project to fetch information about
the project and its environment.

WillDaSilva committed:
Add `MeltanoTracker` class

It will replace the `GoogleAnalyticsTracker`, and add at least a little
bit of abstraction between the places in the codebase using the tracker,
and the backend being used for tracking, which is currently Snowplow.

WillDaSilva committed:
Move `hash_sha256` into `meltano.core.utils`

WillDaSilva committed:
Merge branch 'main' into drop-ga-tracker

WillDaSilva commented on src/meltano/core/tracking/schemas/command_event.json:
```diff
@@ -0,0 +1,73 @@
+            "maxLength": 36
+        },
+        "event": {
+            "description": "",
```
Descriptions are required by SnowcatCloud for most entries. Several are blank because I have yet to fill them in, but others will probably just remain blank as they are only there to satisfy SnowcatCloud's validator.

aaronsteers replied:
Thanks for calling this out. I'm totally fine with leaving these blank for now at least. We can open a new issue to fill those out after the fact, if needed.

WillDaSilva committed:
Add `MeltanoTracker.track_command_event`

WillDaSilva committed:
Move `MeltanoTracker.with_contexts`

WillDaSilva commented:
There are 23 sites in the repository where a `GoogleAnalyticsTracker` method is called, and as best I can tell they will all be replaced by `MeltanoTracker.track_command_event`.

I'm thinking `Project` should have a `tracker` member. This will be the base tracker that only has the environment context. It comes with a context manager `MeltanoTracker.with_contexts` that can be used to track events with additional context. For instance:

```python
project.tracker.track_xyz_event({'name': 'qwerty', 'magnitude': 789})

plugin_context = SelfDescribingJson(
    'iglu:com.meltano/command_event/jsonschema/1-0-0',
    {...}
)

with project.tracker.with_context(plugin_context) as tracker:
    # `tracker` is only provided by the context manager for convenience - the same instance is used    
    assert project.tracker is tracker
    tracker.track_xyz_event({...}) # Tracks an xyz event as before, but this time with additional context
```

WillDaSilva commented:
> I'm thinking `Project` should have a `tracker` member. This will be the base tracker that only has the environment context. 

An alternative would be to decouple the project from the environment context (i.e. split the environment context into an environment context, and a project context), or to start the tracker off with no contexts and make it so that the caller needs to explicitly provide the environment context.

I'm leaning towards the first option, since we'll always want the environment context, and burdening the caller with obtaining it and passing it through every time seems entirely unnecessary.

WillDaSilva committed:
Separate project context from environment context

WillDaSilva committed:
Turn the Snowplow command event into a context

WillDaSilva committed:
Add schema for slim command events

WillDaSilva committed:
Disallow additional properties not specified by the environment or plugins context schemas

WillDaSilva committed:
Remove elements from environment context

aaronsteers commented:
@WillDaSilva - The meeting today was super helpful, I think. Ping me or send me a zoom invite if you need anything further. Thanks!

WillDaSilva committed:
`GoogleAnalyticsTracker` -> `LegacyTracker`

WillDaSilva committed:
Correct DockerHub links

WillDaSilva committed:
Add `is_dev_build` and `is_ci_environment` to Snowplow environment context

WillDaSilva committed:
Add TODO comment to remove `MeltanoTracker.track_struct_event` later

WillDaSilva committed:
Add additional admin to telemetry schema readme

WillDaSilva commented on src/meltano/core/tracking/schemas/README.md:
```diff
@@ -0,0 +1,14 @@
+Login to https://app.snowcatcloud.com/ using your Meltano email address. Request permissions from
+an account admin if necessary. As of the time of writing this, the admins are:
+
+- pat@meltano.com
+- will@meltano.com
+- edgar@meltano.com
+- taylor@meltano.com
+- florian@meltano.com
```
It's satisfying how each name (listed in the order we signed up in) is one character longer than the last.

aaronsteers replied:
We just need to add 'aj@' to the top of the list 😆 

Except I'm not actually an admin.

🤔 ...if only I knew who to ask... 🤔 

Okay, jk. I couldn't resist.

aaronsteers commented on src/meltano/core/tracking/__init__.py:
```diff
@@ -1 +1,172 @@
-from .ga_tracker import GoogleAnalyticsTracker
+"""Meltano telemetry."""
```
I've been trained over time to keep `__init__.py` files as sparse as possible, reserving them mostly just for imports.

Wdyt of moving the bulk of this to something like `/src/meltano/core/tracking/tracker.py`?

WillDaSilva replied:
Will do

aaronsteers commented on src/meltano/core/tracking/schemas/environment_context.json:
```diff
@@ -0,0 +1,184 @@
+            "type": "string",
+            "maxLength": 256
+        },
+        "is_dev_build": {
+            "description": "Whether the installed version of Meltano is a dev build",
+            "type": "boolean"
+        },
```
Can we tell `is_dev_build` with certainty?

aaronsteers replied:
Since I see the marker file is there. I think 'yes'.

WillDaSilva replied:
What we learn is whether the installed version came from `pypi.org` (or e.g. conda-forge), or if it came from someone building/installing from source. Upon reflection, as the PR is currently, it'll also be true for the version of Meltano found in our official Docker images.

aaronsteers replied:
Thanks for the context, @WillDaSilva - For this, and for anything else that's kind of 'nice to have' but causing hangups, I think we can safely trim those from scope if needed and bring back in a future iteration.

aaronsteers commented on src/meltano/core/tracking/environment.py:
```diff
@@ -0,0 +1,124 @@
+        logger.debug(
+            f"Initializing '{type(self).__module__}.{type(self).__qualname__}'"
+        )
+        ci_markers = ('GITHUB_ACTIONS', 'CI')
```
I was going to say to remove this if we don't have confidence, but actually, this looks valid, and valuable.

- https://adamj.eu/tech/2020/03/09/detect-if-your-tests-are-running-on-ci/

aaronsteers commented on src/meltano/core/tracking/environment.py:
```diff
@@ -0,0 +1,124 @@
+                "context_uuid": uuid.uuid4(),
+                "meltano_version": meltano.__version__,
+                "is_dev_build": not release_marker_path.exists(),
+                "is_ci_environment": any(os.environ.get(x, '') == 'true' for x in ci_markers),
```
Do we need to worry about case sensitivity or getting a value `'1'` or `1`?

Maybe this is too clever but not knowing the exact form of the variable's value, something like this seems safer:

```suggestion
                "is_ci_environment": any(str(os.environ.get(x, '')).lower()[:1] in ["1", "t"] for x in ci_markers),
```

WillDaSilva replied:
I originally had pretty much that exact code, but swapped it out for the less-clever version thinking I should stick to something simple that works exclusively with GitLab CI and GitHub Actions. Though in that case the name is no longer accurate. I ought to use `GITLAB_CI` instead of `CI`, and should replace `is_ci_environment` with a string enum called `ci` that can be one of:

- `null`
- `"github_actions"`
- `"gitlab_ci"`

That way we can get more info, and extend it with more CI options by simply adding new enum values to newer versions of the schema.

aaronsteers commented on src/meltano/core/tracking/schemas/environment_context.json:
```diff
@@ -0,0 +1,184 @@
+            ],
+            "maxLength": 256
+        },
+        "node": {
+            "description": "The (potentially not fully-qualified) network name of the computer",
+            "type": [
+                "string",
+                "null"
+            ],
+            "maxLength": 256
+        },
```
Can we leave this and not collect? Seems like tmi.

aaronsteers replied:
Great - thanks very much!

aaronsteers commented on src/meltano/core/tracking/schemas/environment_context.json:
```diff
@@ -0,0 +1,184 @@
+            "minimum": 1,
+            "maximum": 65535
+        },
+        "process_executable_path": {
+            "description": "The path to the executable used to create the process",
+            "type": "string",
+            "maxLength": 4096
+        },
```
Ditto. Seems like TMI and my preference is to not collect.

aaronsteers commented on src/meltano/core/tracking/schemas/environment_context.json:
```diff
@@ -0,0 +1,184 @@
+                "description": "",
+                "type": "object",
+                "properties": {
+                    "process_id": {
+                        "description": "The PID of the process",
+                        "type": "integer",
+                        "multipleOf": 1,
+                        "minimum": 0,
+                        "maximum": 4294967295
+                    },
```
Prefer to not collect.

aaronsteers commented on src/meltano/core/tracking/schemas/environment_context.json:
```diff
@@ -0,0 +1,184 @@
+                        "minimum": 0,
+                        "maximum": 4294967295
+                    },
+                    "process_name": {
+                        "description": "The name of the process",
+                        "type": "string",
+                        "maxLength": 1024
+                    },
```
Preference to hash or don't collect.

aaronsteers reviewed (dismissed):
More feedback coming but this is first pass.

WillDaSilva commented:
GitLab CI for this branch is currently failing hard because the Poetry lock file is out of sync with `pyproject.toml`.

It's out of sync because I added a couple of dependencies (`backports.cached_property` and `tzlocal`), but I haven't been able to get Poetry to update the lock file. Poetry just keeps taking more and more memory before ultimately getting killed by the OOM killer.

I've verified that no versions of those dependencies were installed by Meltano previously, and I've manually checked the transitive dependencies of the 2 dependencies I added. I don't see any conflicts, so I'm not sure what's going wrong here.

pandemicsyn commented:
Not sure why but I think its hanging on backports.cached-property:

```
$ poetry lock -vvv
<-- spam for infinity --->
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
   1: selecting backports.cached-property (1.0.1)
^C   1: Version solving took 12.258 seconds.
   1: Tried 1 solutions.
```

WillDaSilva commented:
@pandemicsyn I'll switch to https://pypi.org/project/cached-property/ and see if Poetry likes that better.

WillDaSilva commented:
@pandemicsyn Switching to the other cached property package seems to have resolve the issue. I've used both in the past without much trouble, and we're only going to be using the basic features of it, and it's temporary until we drop Python 3.7 support, so I think we're fine with taking the fancier one for now that is more than just a backport.

WillDaSilva committed:
Switch cached property dependency & update Poetry lock

pandemicsyn commented:
> so I think we're fine with taking the fancier one for now that is more than just a backport.

@WillDaSilva 👍 sounds good.

fwiw - its weird - has to be a conflict caused by another dep because this minimal pyproject.toml in a fresh project generates a lock file fine 😕 

```yaml
[tool.poetry]
name = "test"
version = "0.1.0"
description = ""
authors = ["pandemicsyn <syn@ronin.io>"]

[tool.poetry.dependencies]
python = "3.7,<3.10"
"backports.cached-property" = "^1.0.1"
meltano = "^1.105.0"

[tool.poetry.dev-dependencies]

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

WillDaSilva committed:
Move tracker out of `__init__.py`

WillDaSilva commented on src/meltano/core/tracking/environment.py:
```diff
@@ -0,0 +1,124 @@
+                "python_version": platform.python_version(),
+                "python_implementation": platform.python_implementation(),
+            }
+            | self.system_info
+            | self.process_info,
```
Good catch - it is not. I'll update it to use `**` unpacking to merge them instead.

WillDaSilva committed:
Address misc review comments

WillDaSilva committed:
Do not collect platform node

WillDaSilva committed:
Merge branch 'main' into drop-ga-tracker

WillDaSilva commented:
Going back to our discussion about having a plugin event versus a plugins context, I'm thinking we should have a plugin context that is for a single plugin. That context is entered when the plugin in loaded. We'll then have a slim `plugin_loaded` event that is emitted within that context.

Though I have not used Snowplow before, I get the impression that it's ideal for us to rely primarily on contexts for gathering information, along with slim events whose main purpose is to provide info on what contexts were activate when the event was emitted.

Maybe I'm totally off-base with how Snowplow ought to be used. What do you all think? @aaronsteers @pnadolny13 @tayloramurphy @pandemicsyn

pnadolny13 commented:
> Maybe I'm totally off-base with how Snowplow ought to be used. What do you all think? @aaronsteers @pnadolny13 @tayloramurphy @pandemicsyn

@WillDaSilva yeah that makes sense to me. I personally havent used Snowplow until now and its super flexible on how we can send data so its really up to us.

The only thing I'd add is that as part of a modeling discussion with Joao from Snowcat Cloud he was encouraging us to model the data in a way that encapsulates all our data in a few flexible and generally useful events/context (i.e. Command vs Meltano ELT or Meltano Run), each context and event has its own schema so making them generic reduces the need to constantly be updating the schemas or adding new ones. I think thats normal modeling/DRY principles anyways but still worth sharing.

edgarrmondragon commented:
> I'm not sure we wanna fire an event for every plugin seen/loaded

I agree with @pandemicsyn. There some inefficiencies in plugin loading a.t.m. (e.g. extra roundtrips) so events would be overcounted there.

Commit:
Merge branch 'main' into drop-ga-tracker

pandemicsyn commented on src/meltano/core/tracking/schemas/project_context.json:
```diff
@@ -0,0 +1,59 @@
+            "maxLength": 36
+        },
+        "environment_name": {
+            "description": "The name of the active environment for the Meltano project",
```
We'll want to hash this one.

edgarrmondragon commented on src/meltano/core/tracking/environment.py:
```diff
@@ -0,0 +1,125 @@
+                GetCurrentProcess(), ctypes.byref(mask), ctypes.byref(DWORD_PTR())
+            ):
+                return bin(mask.value).count("1")
+            logger.debug("Call to 'GetProcessAffinityMask' failed")
+        logger.debug(
+            "Unable to determine the number of available CPU cores; "
+            "falling back to the total number of CPU cores"
+        )
```
@WillDaSilva would these messages be useful for end users debugging their Meltano deployment? Might be for us in managed service setting, but then telemetry is not as useful 😅

pandemicsyn commented on src/meltano/core/tracking/schemas/plugins_context.json:
```diff
@@ -0,0 +1,117 @@
+                            "trasformer"
+                        ]
+                    },
+                    "type": {
```
drop "type"

pandemicsyn commented on src/meltano/core/tracking/schemas/plugins_context.json:
```diff
@@ -0,0 +1,117 @@
+                        "maxLength": 256
+                    },
+                    "name_hash": {
+                        "description": "",
```
From zoom: desc should describe "this is a 64char hash field" , but drop pattern, max, min constraints.

pandemicsyn commented on src/meltano/core/tracking/schemas/plugins_context.json:
```diff
@@ -0,0 +1,117 @@
+                        "maxLength": 64,
+                        "minLength": 64
+                    },
+                    "version": {
```
From zoom: Drop version (not present in plugin today)

pandemicsyn commented on src/meltano/core/tracking/schemas/plugins_context.json:
```diff
@@ -0,0 +1,117 @@
+                        "type": "string",
+                        "maxLength": 256
+                    },
+                    "plugin_definition_hash": {
```
From zoom: renamed to plugin_lock_hash if accessible. make optional.

pandemicsyn commented on src/meltano/core/tracking/schemas/plugins_context.json:
```diff
@@ -0,0 +1,117 @@
+                        "maxLength": 64,
+                        "minLength": 64
+                    },
+                    "containerized": {
```
From zoom: drop, should be invoker field on command event.

pandemicsyn commented:
Ok @aaronsteers @pnadolny13 and I just a had a group PR review over zoom to go over the schema's. Logged all the issues inline above. Next steps are:

1. resolve the circular import
2. address the PR feedback that came on the of the call
3. add cli tracking events to `meltano invoke`
4. merge this pr
5. profit!

pandemicsyn commented on src/meltano/core/tracking/schemas/plugins_context.json:
```diff
@@ -0,0 +1,117 @@
+                    "category": {
+                        "description": "",
+                        "type": "string",
+                        "enum": [
```
From zoom: drop enum, either use string or https://json-schema.org/draft/2020-12/json-schema-validation.html#rfc.section.9.5 (/cc @aaronsteers )

Commit:
Move legacy tracker to its own package

Commit:
flak8 fixes

pandemicsyn commented:
This is gonna probably have *a lot* of linter & doc convention failures (both for the new additions and because theres adjecent updates in some files that haven't been touched during other linter fixups).

Commit:
Hash env name in project context

Commit:
Update schema's from PR feedback

Commit:
Restructure on disk layout of schema's to play nice with igluctl

Commit:
Quick test to make sure initializing the tracker doesn't kaboom

Commit:
Maybe fix tests

pandemicsyn commented on src/meltano/core/tracking/environment.py:
```diff
@@ -0,0 +1,120 @@
+from __future__ import annotations
+
+import ctypes
+import ctypes.wintypes
```
This seems to not play nice with python 3.7: https://github.com/meltano/meltano/runs/6734175975?check_suite_focus=true , but seems to at least pass tests on 3.8+. Do we need to ensure that this only imports on Windows platforms?

The python docs don't explicitly mention that this should fail on non-windows systems but various posts like https://stackoverflow.com/questions/20897842/error-in-importing-ctypes-wintypes imply just that.

We're currently using this to get a core count on windows - so we could omit it if we really don't actually need the core count anyway.

pandemicsyn replied:
Update. Ended up just nuking it because I wasn't sure how fragile ctypes where and we don't really test on windows (yet). Happy to put it back in (wrapping the import in try/except) though if we want it.

aaronsteers replied:
@pandemicsyn - Yes, sounds like that is the right call. Sorry I didn't see this earlier.

Commit:
Linter fixes and removal of win specific core discovery

Commit:
Test meltano invoke implementation and more linting

Commit:
Fix 3.7 test failure due to mock import rewrites

Commit:
snowplow-micro support

pandemicsyn commented on .pre-commit-config.yaml:
```diff
@@ -59,5 +59,6 @@ repos:
     args:
     - "--py3-plus"
     - "--py36-plus"
+    - "--keep-mock"
```
first time running into this but ended up having to add this because of https://github.com/asottile/pyupgrade/issues/314 (just didn't feel like i had the cycles to rewrite tests)

pandemicsyn commented on src/meltano/core/tracking/tracker.py:
```diff
@@ -0,0 +1,196 @@
+                label=self.project_id,
+            )
+        except Exception as ex:
+            logger.warning(
```
@aaronsteers I think im gonna switch these to logger.debug. If we run into issues with telemetry I don't wanna spam the end-user with errors, and we can easily enough troubleshoot using --log-level=debug. wdyt ?

aaronsteers replied:
Yeah - makes sense! 👍

aaronsteers commented on src/meltano/core/tracking/tracker.py:
```diff
@@ -0,0 +1,196 @@
+        """
+        self.track_unstruct_event(
+            SelfDescribingJson(
+                "iglu:com.meltano/cli_event/jsonschema/1-0-0", event_json
```
Here, and everywhere we have something like `"iglu:.../1-0-0"`, what do you think of moving these to constant string declarations at the top of the page.

As a single constant it might be something like `COMMAND_EVENT_SCHEMA_IGLU`.

Or as a two-part, it might be something like `COMMAND_EVENT_SCHEMA_IGLU` and `COMMAND_EVENT_SCHEMA_IGLU_VERSION`

WillDaSilva replied:
A constant dictionary mapping schema names to their version might make sense for this.

aaronsteers commented on src/meltano/core/utils/__init__.py:
```diff
@@ -421,3 +422,15 @@ def human_size(num, suffix="B"):
     return f"{val:3.1f}{prefix}{suffix}"
+
+
+def hash_sha256(value: str) -> str:
+    """Get the sha256 hash of a string.
+
+    Args:
+        value: the string value to hash.
+
+    Returns:
+        The hashed value of the given string.
+    """
+    return hashlib.sha256(value.encode()).hexdigest()
```
Great call to move this into `utils`. 👍

aaronsteers commented on .gitlab/ci/publish.gitlab-ci.yml:
```diff
@@ -7,14 +7,25 @@ publish:
+    - ARCHIVE="$(sh -c 'echo "$1"' sh meltano*.tar.gz)"
+    - DIR="$(basename $ARCHIVE .tar.gz)"
+    - tar -xvzf "$ARCHIVE"
+    - touch "$DIR/src/meltano/core/tracking/.release_marker"
```
Python packaging is weird sometimes about what actually ships and what doesn't. Can we double check this file gets bundled and installed with the operation? (Either in docs or through a 'real' test.)

Also - and I just realized this now - I think this breaks if a user does `pipx install git+https://github.com/meltano/meltano.git`

For now, I think this can be a bug we would live with, but perhaps we open an issue to track/prioritize in a future PR. 

Wdyt?

pandemicsyn replied:
👍 I'll log an issue, im not as well versed with python packaging as y'all so gonna defer to y'all on how realiable this is.

pandemicsyn replied:
https://github.com/meltano/meltano/issues/6021

aaronsteers reviewed (approved):
I did a deep review of each file - minus the micro.conf file which I skimmed.

While I lost my original 15 other comments due to user error (🤦 ) , most were cosmetic and general feedback. 

I will make one more pass to see if I can recall/repost some of my other comments, but none of those were blocking issues.

@pandemicsyn and @WillDaSilva - Thank you both for your hard work on this, @WillDaSilva for starting it and @pandemicsyn for getting us to the finish line.

@pandemicsyn - You are clear to merge when ready.

aaronsteers commented on src/meltano/core/tracking/README.md:
```diff
@@ -0,0 +1,59 @@
+When a new version of the schema is introduced , it must also be updated in our schema registry,
+SnowcatCloud.
+
+Login to https://app.snowcatcloud.com/ using your Meltano email address. Request permissions from
+an account admin if necessary. As of the time of writing this, the admins are:
+
+- pat@meltano.com
+- will@meltano.com
+- edgar@meltano.com
+- taylor@meltano.com
+- florian@meltano.com
```
I've got admin too now 😄 

```suggestion
Login to https://app.snowcatcloud.com/ using your Meltano email address. Request permissions from
an account admin if necessary. As of the time of writing this, the admins are:

- aj@meltano.com
- pat@meltano.com
- will@meltano.com
- edgar@meltano.com
- taylor@meltano.com
- florian@meltano.com
```

aaronsteers commented on src/meltano/core/tracking/README.md:
````diff
@@ -0,0 +1,59 @@
+- taylor@meltano.com
+- florian@meltano.com
+
+Once SnowcatCloud provides an API to update the schemas, we should add a CI job to update them
+automatically when a release pipeline is run with changes in this directory.
+
+To test/validate schemas [snowcloud-micro](https://github.com/snowplow-incubator/snowplow-micro) can be used:
+
+
+```
+docker run \
+  --mount type=bind,source=$(pwd),destination=/config \
+  -p 9090:9090 \
+  snowplow/snowplow-micro:1.2.1 \
+  --collector-config /config/micro.conf \
+  --iglu /config/iglu.json
+```
````
Really appreciate these instructions being inline and included in the repo. Will be a huge help.

Commit:
Fires a known good event!

aaronsteers commented on src/meltano/core/tracking/environment.py:
```diff
@@ -0,0 +1,139 @@
+            return {
+                "num_cpu_cores": psutil.cpu_count(),
+                "num_cpu_cores_available": self.num_available_cores,
+                "process_executable_path": process.exe() or None,
```
@pandemicsyn called this out to me - sorry I missed it earlier.

`process_executable_path` can _definitely_ contain PII and is not permittable to be included.

Even this is PII:

`/usr/ajsteers/source/meltano`

And in addition to PII, there could be any amount of sensitive/non-public info in the path:

`/usr/ajsteers/projects/mysecretworldtakover/meltano/clone-army-analytics/`

aaronsteers reviewed (approved):
I'm approving (again) but the `process_executable_path` definitely should be removed. As noted.

pandemicsyn commented:
It lives!

![Screen Shot 2022-06-04 at 2 36 36 PM](https://user-images.githubusercontent.com/333354/172023106-331d4652-f7c1-4a26-8d10-58f116bed9a2.png)

aaronsteers commented on src/meltano/core/tracking/iglu-client-embedded/schemas/com.meltano/environment_context/jsonschema/1-0-0:
```diff
@@ -0,0 +1,169 @@
+{
+    "$schema": "http://iglucentral.com/schemas/com.snowplowanalytics.self-desc/schema/jsonschema/1-0-0#",
+    "description": "Schema for a Snowplow context describing the environment in which Meltano is operating",
```
Total nit, but a slight tweak here makes it more clear we aren't talking about the 'environments' feature.

```suggestion
    "description": "Schema for a Snowplow context describing the environment in which the Meltano process is executing",
```

aaronsteers commented:
Done. 
![image](https://user-images.githubusercontent.com/18150651/172023180-ec530780-6503-4463-9505-b8811a55c721.png)

@pandemicsyn - I went through all the files where my prior comments were lost. You should have all my feedback now.

Also - in case you haven't already seen, there are some merge conflicts noted that are pending resolution:

![image](https://user-images.githubusercontent.com/18150651/172023196-c0c297aa-34fd-485a-b6c5-71c841a0e998.png)

Commit:
Update src/meltano/core/tracking/README.md

Co-authored-by: Aaron ("AJ") Steers <aj@meltano.com>

Commit:
Update src/meltano/core/tracking/iglu-client-embedded/schemas/com.meltano/environment_context/jsonschema/1-0-0

Co-authored-by: Aaron ("AJ") Steers <aj@meltano.com>

Commit:
Merge branch 'main' into drop-ga-tracker

Commit:
Verify legacy tracking events fire

pandemicsyn commented:
Ok, with the latest change set the legacy tracking is working as well. Verified that the events are at least accepted as "good" by snowplow micro (can't speak to whether they look like they old ones 😁 ).

Commit:
Last PR review feedback
